rfcs/rdm-0064-custom-fields.md

```diff
@@ -0,0 +1,440 @@
+
+## Detailed design
+
+An instance administrator can add custom fields. **Updating and deleting a custom field is not possible**. This constraint is imposed to avoid data migration and potential inconsistencies.
+
+### Configuring a _custom field_
+
+The first step would be to define how custom fields can be added/configured. A top to bottom approach (from more to less abstracted) will be followed to explain the configuration. Taking the following use case:
+
+_At CERN you want to add the experiment to which the record belong to, assuming experiments come from a controlled vocabulary. In addition, I want to be able to filter the record search by experiment._
+
+The instance administrator would add it in the `invenio.cfg` configuration file, under `RDM_RECORDS_CUSTOM_FIELDS` and `RDM_RECORDS_CUSTOM_FIELDS_UI`. The following example shows only the configuration needed to **enable the custom field** in the data model not how to display them. This second part will be explained in the [_Adding-a-custom-UI-widget_](#Add a custom UI widget) section.
+
+```python
+RDM_RECORDS_CUSTOM_FIELDS = {
+    "experiment": VocabularyCF(
+        name="experiment",
+        vocabulary_id="cernexperiments",
+    ),
+}
+```
+
+> Note: In the above configuration and along this RFC configuration variables are prefixed with `RDM_RECORDS_`, this means that custom fields are being added to _records_. However, the same configuration variables are available in communities and will be used then adding custom fields to them. Namely, `COMMUNITIES_CUSTOM_FIELDS` and `COMMUNITIES_CUSTOM_FIELDS_UI`.
+
+Custom fields configuration, abstracts the data model in `<Type>CF` classes. These will be in charge or returning the Elasticsearch mapping and marshmallow schema that will be applied to the field. In the previous example `VocabularyCF`.
+
+```python
+class BaseCF(ABC):
+"""Base Custom Field class."""
+
+    def __init__(self, name):
+        """Constructor."""
+        super().__init__()
+        self.name = name
+
+    @property
+    @abstractmethod
+    def mapping(self):
+        """Return the mapping."""
+        pass
+
+    @property
+    @abstractmethod
+    def schema(self):
+        """Marshmallow schema for custom fields."""
+        pass
+
+    @property
+    def ui_schema(self):
+        """Marshmallow UI schema for custom fields.
+
+        Controls how the field will be serialized in the UI when using the
+        `UIJSONSerializer`. If not overriden, the default implementation will use
+        the `BaseCF.schema` property.
+        """
+        return self.schema
+
+class TextCF(BaseCF):
+"""Text custom field."""
+
+    @property
+    def mapping(self):
+        return {"type": "text"}
+
+    @property
+    def schema(self):
+        return SanitizedUnicode()
+```
+
+Note that all custom fields will be added inside the `custom` parent field, at the same level than `metadata`.
```

Comment:
Did we assess all the implications (if any!) of having the `custom` outside the `metadata` field?

Reply:
I'm not sure. Having it inside `metadata` poses very big implementation challenges. And from the top of my mind I cannot think of any big inconvenience having it outside. However, we might want to sit down and think about it before going further along. You're right!